### PR TITLE
WIP: Refactoring of quest list using bootstrap-table

### PR DESCRIPTION
**Pull request by Enprogames**
### What?
I've made a first attempt at refactoring the quest list to be a bootstrap-table.

View these changes by pulling this branch and navigating to http://hackerspace.localhost:8000/quests/available2/.

### Why?
Once this is working properly, it will allow more dynamic actions to be performed on the lists of quests, such as sorting, filtering, etc. See #1369.
### How?
I updated `quests2.html` to match `quests.html`, and made it point to `tab_quests_available2.html`. This gives this code a place to live while it is still a work in progress.

Most of the work on this issue was done inside of `tab_quests_available2.html`. The code mostly matches [this JSFiddle Example](https://jsfiddle.net/enprogames/cv8f2Lgh/36/), but with additional complexity. So essentially, JavaScript code was added to do the following:
- Initializing the Bootstrap Table
- Adding the content to the hidden divs
- Expanding the row when it is clicked, and collapsing any other expanded rows
- Collapsing the row when it is clicked again

#### ❗ Updates ❗ 
- I moved the CSS and JavaScript code for this to `static/quest_manager/css/bootstrap-table-accordian.css` and `static/quest_manager/js/bootstrap-table-accordian.js`. They are imported in `quest_manager/base.html` and `quest_manager/quests2.html`.
- I fiddled around with the CSS a bit and got it looking quite similar to the old version.
- To fix the problem of the detailed content not being updated after being loaded in by the `ajax_content_loading.html` code, I used something called a "Mutation Observer". It listens for changes to the detailed view content and updates it if it is changed. From my research, it looked like this feature doesn't necessarily work in internet explorer, so we might want to consider a different solution.

<details>

  <summary>Click to expand code</summary>

```jsx
$table.on('expand-row.bs.table', function (e, index, row, $detail) {
      var row_html_id = row._id;
      var quest_id = parseInt(row_html_id.match(/\d+$/)[0], 10);

      $detail.html('<div class="detail-container" style="display:none;"></div>');
      var $detailContainer = $detail.find('div');

      // Create a function to update the row's content
      var updateContent = function () {
        var detailContent = $("#collapse-quest-" + quest_id).html();
        $detailContainer.html(detailContent);
      };

      // Update the row's content initially
      updateContent();

      // Create an observer that updates the row's content when the span's content changes
      var observer = new MutationObserver(updateContent);
      observer.observe($("#collapse-quest-" + quest_id)[0], { childList: true, subtree: true });

      // When the row is expanded, slide down the detail container
      $detailContainer.slideDown(collapseFinished);
    });
```
</details>

### Testing?
No unit tests have been added for this at this point.

These changes have been tested manually for student and admin users. I tested in Firefox.

These changes can be seen by logging in as a student and going to the URL http://hackerspace.localhost:8000/quests/available2/.
### Screenshots (if front end is affected)
image

### Anything Else?
The only problems that I can still see are to do with styling:
- The CSS styling still has lots of room for improvement.
- ❗For some reason, they bootstrap-table rows are a bit smaller than the original ones were. I will need time to look into the styling for this.
- ~~There is a strange issue that when certain rows are collapsed, some of the table row content is shifted around slightly. The content within the detailed views is likely causing some styling rules to be activated or deactivated unexpectedly.~~

#### ❗ Ideas
- We might be able to simplify our code if we go for a more vanilla version of this functionality e.g. without the expanding/collapsing animations. It looks like bootstrap-table provides some better ways of passing in table data than how it's being done in my code, such as by using the `detailFormatter` function to insert content from an API endpoint rather than from hidden <span> elements.
    - This example provides most of the same functionality of expanding/collapsing tables, but with much less code: https://examples.bootstrap-table.com/index.html?bootstrap3#methods/expand-collapse-row.html 
- I've also refactored the `tab_quests_submission.html`, `tab_quests_approvals.html` and `category_detail_content.html` files, and I've noticed that their quest lists are all extremely similar, only varying slightly. I think it might be a good idea to consolidate all of the accordian quest list functionality into one template.

### Review request
@tylerecouture

### CODE REVIEW

File: src/quest_manager/templates/quest_manager/tab_quests_available2.html [hunk 1]
```diff
@@ -4,83 +4,170 @@
...
+    - Expanding the row when it is clicked
+    - Collapsing the row when it is clicked again
+
+    Somewhere else (where?), there is a block of javascript that is responsible for:
+    - Adding "active" and "collapse" classes to rows when they are expanded or collapsed
+
+{% endcomment %}
+
+<!-- Import Bootstrap Table JS -->
+<script src="https://cdnjs.cloudflare.com/ajax/libs/bootstrap-table/1.18.3/bootstrap-table.min.js"></script>
```
[the remark this answers is not in the document]
Comment (tylerecouture) [reply src/quest_manager/templates/quest_manager/tab_quests_available2.html:126]: @Enprogames See if you can add the full integrity check for this js import

**Comment by tylerecouture** [comment]
Code is broken, so I can't take a closer look, but some initial comments:

> Currently, all of the CSS and JavaScript code for this exists inside of tab_quests_available2.html. This is because it made it easier to work on, and because I wasn't sure which other files to put these in. This should be fixed soon.

I would put these into their own `bootstrap-table-accordian.css|.js` files within `quest_manager/static/css|js`

> Expanded quests don’t have blue highlight. These CSS rules aren’t being applied properly.

> The quest list table doesn’t have the same styling as the old version. The styling here has several moving parts and is quite complex, so making it look like the old one would be a big undertaking.

Styling doesn't have to match exactly, it just needs to look clean, and good.  Start by giving the table element the classes .table and .table-hover

See https://getbootstrap.com/docs/3.3/css/#tables

>Upon expanding the first quest, the loading prompt doesn’t go away unless you collapse/expand it again. It looks like the HTML content isn’t being inserted and updated as it should be?

See `ajax_content_loading.html`

> I can't figure out where the JavaScript code exists that was making the old collapsing functionality work, but it is still affecting these changes in a wonky way. So either this new code or that old code will need to be refactored.

Current accordions use Bootstrap built in accordion that is triggered by css classes and ids. See here:
https://getbootstrap.com/docs/3.3/javascript/#collapse-example-accordion

**Review by tylerecouture — COMMENTED** [review]
Great progress.  Request my review again when you'd like me to take another look.

**Review by tylerecouture — CHANGES_REQUESTED** [review]
As discussed.

**Review by tylerecouture — CHANGES_REQUESTED** [review]
Haven't dug into the code yet, just played around with the changes in a browser, and it is looking awesome!!  A few things to tweak:
1.  Mobile view is messed up, I think because the table headings aren't set to disappear, only the table columns, so left with some broken layout in mobile/narrow screen:
image

2. The text is smaller (e.g. 14pt  vs 16pt in the old version for the quest title)

3. Column alignment is funky, compared to old version (I know the sorting arrows mess it all up a bit, but maybe align the headings to right and align the XP and campaign content to center?  Try a few things to make it look better:
image

4. Add the Search!  This should just be a simple data attribute in the bootstrap-table, see the students or tags list tables for an example. `data-search='true'`

**Comment by tylerecouture** [comment]
@Enprogames Also, please make this change: replace the expiry and last column (where the status icons appear) with a single column: Tags (which should use the `{{config.custom_name_for_tag }}` or whatever that field is called.

Then, for the column content for each quest, put a comma separate list of tags, I think `{{q.tags.all|join:", "}}`

You can also now replace the old code with your new stuff (don't delete the old ones yet though, please just rename them all with suffix _old and maybee provide access to it via url `/available/old` or something like that

**Comment by Enprogames** [comment]
> @Enprogames Also, please make this change: replace the expiry column and all its logic witha column to show tags instead: Tags (which should use the `{{config.custom_name_for_tag }}` or whatever that field is called).
> 
> Then, for the column content for each quest, put a comma separate list of tags, I think `{{q.tags.all|join:", "}}`
> 
> You can also now replace the old code with your new stuff (don't delete the old ones yet though, please just rename them all with suffix _old and maybee provide access to it via url `/available/old` or something like that

- [x] Add search field
- [x] Increase text size to 16px
- [x] Replace expiry column with custom tags
- [x] Make this the official quest_available view
- [ ] Fix column alignment
- [ ] Fix broken mobile view:
    - This is proving to be quite difficult. The bootstrap and bootstrap-table documentations don't seem to have any good instructions on this. [It seems like the only straightforward way is to use `@media` CSS tags](https://stackoverflow.com/a/42534400/6946463)? I will try this soon.

**Pull request merged**